### PR TITLE
#162725235 Users should bookmark articles

### PR DESCRIPTION
**Pull request by armstrongsouljah**
#### What does this PR do?
- Enables users to bookmark articles for reading later

#### Description the task to be implemented
- Users should be able to bookmark articles for reading later via a bookmark tab, button or link
- Implement bookmarking functionality
- Write tests for bookmarking functionality
- Render a list of bookmarked articles in the user's profile

#### How can this be manually tested?
- Fetch this branch by running `git fetch origin ft-bookmark-articles-162725235`
- Checkout to this branch by running ` git checkout ft-bookmark-articles-162725235`
- Install dependencies by running `yarn install`
- Start the local server by running `yarn start`
- Log in with either `social auth` or the `login form`
- Bookmark any article by clicking on the bookmark icon
- View your bookmarks by visiting your profile and checking the bookmarked articles tab.

#### Screenshots if necessary.
- Article bookmarked
screenshot 2019-03-05 at 17 47 22
- Article unbookmarked
screenshot 2019-03-05 at 17 47 12

#### What are the related pivotal stories?
[#162725235](https://www.pivotaltracker.com/story/show/162725235)

### CODE REVIEW

**Comment by RrNn** [comment]
@armstrongsouljah could you give the `bookmark` icon a better cursor on hover.

**Comment by armstrongsouljah** [comment]
> @armstrongsouljah could you give the `bookmark` icon a better cursor on hover.

okay.

**Comment by armstrongsouljah** [comment]
@RrNn I have added the hover style. All is good to go.

**Review by MuhweziDeo — DISMISSED** [review]
LGTM

**Review by calvinpete — COMMENTED** [review]
@armstrongsouljah Please refactor the command line for installing dependencies to `yarn install`.

**Review by kdaglas — DISMISSED** [review]
i have tested it, i like the tooltips implementation

**Comment by MuhweziDeo** [comment]
Please @armstrongsouljah  check you have a bug, when I bookmark an article and visit the others it shows that I booked marked them cc @calvinpete

**Comment by armstrongsouljah** [comment]
Okay @calvinpete

**Review by calvinpete — CHANGES_REQUESTED** [review]
@armstrongsouljah It seems after bookmarking one article, the bookmark icon for other articles also changes to the bookmarked status color

File: tests/ArticleReadView.test.js [hunk 1]
```diff
@@ -1,21 +1,30 @@
 import React from 'react';
 import { shallow } from 'enzyme';
 import { ArticleReadView, mapStateToProps } from '../src/views/ArticleReadView';
-
+import { toast } from 'react-toastify';
```
[the remark this answers is not in the document]
Comment (malfahad) [reply tests/ArticleReadView.test.js:3]: @armstrongsouljah  please deal with the hound violations

**Comment by malfahad** [comment]
@armstrongsouljah please deal with the hound violations

**Review by malfahad — DISMISSED** [review]
LGTM

**Review by MuhweziDeo — DISMISSED** [review]
LGTM

**Review by calvinpete — DISMISSED** [review]

File: src/reducers/__tests__/bookmarkListReducer.test.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      type: ActionTypes.BOOKMARK_LIST_FETCHED,
+      payload: [{ title: 'dkekkd', slug: 'dededed' }],
+    };
+    const expectedState = {
+      bookmarks: [{ title: 'dkekkd', slug: 'dededed' }],
+      bookmarks_empty_message: '',
+      isBookmarked: false,
+
+    };
+    // Array.prototype.filter =  jest.fn();
```
Comment (RrNn) [inline src/reducers/__tests__/bookmarkListReducer.test.js]: Remove commented code.

File: src/reducers/__tests__/bookmarkListReducer.test.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      type: ActionTypes.BOOKMARK_SUCCESS,
+      bookmarks: [],
+    };
+    const expectedState = {
+      bookmarks: [],
+      bookmarks_empty_message: '',
+      isBookmarked: true,
+
+    };
+    // Array.prototype.filter =  jest.fn();
```
Comment (RrNn) [inline src/reducers/__tests__/bookmarkListReducer.test.js]: Plus this

File: src/reducers/__tests__/bookmarkListReducer.test.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      type: ActionTypes.UNBOOKMARK_SUCCESS,
+      bookmarks: [],
+    };
+    const expectedState = {
+      bookmarks: [],
+      bookmarks_empty_message: '',
+      isBookmarked: false,
+
+    };
+    // Array.prototype.filter =  jest.fn();
```
Comment (RrNn) [inline src/reducers/__tests__/bookmarkListReducer.test.js]: Plus this

File: src/reducers/__tests__/bookmarkListReducer.test.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    const action = {
+      type: '',
+      bookmarks: [],
+    };
+    const expectedState = {
+      bookmarks: [],
+      isBookmarked: undefined,
+      bookmarks_empty_message: '',
+    };
+    // Array.prototype.filter =  jest.fn();
```
Comment (RrNn) [inline src/reducers/__tests__/bookmarkListReducer.test.js]: Plus this

**Review by kdaglas — DISMISSED** [review]
i have tested it and the bug was fixed... LGTM

**Review by calvinpete — APPROVED** [review]
LGTM

**Comment by RrNn** [comment]
I'll go ahead and merge in this, but we shall need to fix our tests and remove the errors.

**Comment by armstrongsouljah** [comment]
Okay.

**Comment by RrNn** [comment]
We also have an unclear test which has an uncalled mock function. Run the current develop tests to see it.

**Comment by armstrongsouljah** [comment]
We'll look into those in the morning.

**Comment by kdaglas** [comment]
I actually saw that uncalled mock function and I also showed it to @malfahad, we will look at it today

**Pull request merged**